add docs for the Navigation API (#21960)
mdn/content: files/en-us/web/api/navigationhistoryentry/index.md

---
title: NavigationHistoryEntry
slug: Web/API/NavigationHistoryEntry
page-type: web-api-interface
tags:
  - API
  - Experimental
  - History
  - Interface
  - Landing
  - Navigate
  - NavigationHistoryEntry
  - Navigation API
  - Reference
  - Scroll
  - Traversal
browser-compat: api.NavigationHistoryEntry
---

{{APIRef("Navigation API")}}{{seecompattable}}

The **`NavigationHistoryEntry`** interface of the {{domxref("Navigation API", "Navigation API", "", "nocode")}} represents a single navigation history entry.

These objects are commonly accessed via the {{domxref("Navigation.currentEntry")}} property and {{domxref("Navigation.entries()")}} method.

> The Navigation API only exposes history entries created directly by the application (i.e. not {{htmlelement("iframe")}} navigations or cross-origin navigations), providing an accurate list of all previous history entries just for your app. This makes traversing the history a much less fragile proposition than the older {{domxref("History API", "History API", "", "nocode")}}.

Author: Good call; updated in the next commit.

And don't worry about the speed of review — I appreciate that this is a really big one.

{{InheritanceDiagram}}

## Instance properties

_Inherits properties from its parent, {{DOMxRef("EventTarget")}}._

- {{domxref("NavigationHistoryEntry.id", "id")}} {{ReadOnlyInline}}
  - : Returns the `id` of the history entry. This is a unique, UA-generated value that always represents the history entry, useful to correlate a history entry with an external resource such as a storage cache.
- {{domxref("NavigationHistoryEntry.index", "index")}} {{ReadOnlyInline}}

>   - : Returns the index of the history entry in the history entries list (i.e. returned by {{domxref("Navigation.entries()")}}), or `-1` if the entry does not appear in the list.

Reviewer: ```suggestion
  - : Returns the index of the history entry in the history entries list (that is, the list returned by {{domxref("Navigation.entries()")}}), or `-1` if the entry does not appear in the list.
```

Author: changed

  - : Returns the `key` of the history entry. This is a unique, UA-generated value that represents the history entry's slot in the history entries list, used to navigate to this place in the history via {{domxref("Navigation.traverseTo()")}}. It will be reused by other entries that replace the entry in the list (i.e. if the {{domxref("NavigateEvent.navigationType")}} is `replace`).
- {{domxref("NavigationHistoryEntry.sameDocument", "sameDocument")}} {{ReadOnlyInline}}
  - : Returns `true` if this history entry is for the same `document` as the current {{domxref("Document")}} value, or `false` otherwise.
- {{domxref("NavigationHistoryEntry.url", "url")}} {{ReadOnlyInline}}
  - : Returns the absolute URL of this history entry.

## Instance methods

_Inherits methods from its parent, {{DOMxRef("EventTarget")}}._

- {{domxref("NavigationHistoryEntry.getState", "getState()")}}
  - : Returns a clone of the available state associated with this history entry.

## Events

- {{domxref("NavigationHistoryEntry/dispose_event", "dispose")}}
  - : Fires when the entry is no longer part of the history entry list.

## Examples

```js

> async function initHomeBtn() {

Author: Good point. I've updated those snippets

  // Get the key of the first loaded page
  // so the user can always go back there.
  const {key} = navigation.currentEntry;
  backToHomeButton.onclick = () => navigation.traverseTo(key);
}

async function handleNavigate(url) {
  // Navigate away, but the button will always work.
  await navigation.navigate(url).finished;

> }

Author: This is a nice little example, but I kinda think it is overkill for what I'm trying to show here. I've assumed we are dealing with an SPA, as this is what this API is for. So realistically you would be dealing with the same page all the time, just with different content in injected into the main content container. The idea is that you are getting the key of the currentEntry when the current entry is the home page, and setting up the event listener so that the home button will always navigate back to that history entry.

I've updated the code comments to make them less confusing ("different page" did imply that we are loading a different page ;-) ), and I've added a skeleton navigation handler to indicate further that these are single-page navigations.

  - : Returns the `key` of the history entry. This is a unique, UA-generated value that represents the history entry's slot in the entries list rather than the entry itself. It is used to navigate that particular slot via {{domxref("Navigation.traverseTo()")}}. The `key` will be reused by other entries that replace the entry in the list (that is, if the {{domxref("NavigateEvent.navigationType")}} is `replace`).
- {{domxref("NavigationHistoryEntry.sameDocument", "sameDocument")}} {{ReadOnlyInline}}
  - : Returns `true` if this history entry is for the same `document` as the current {{domxref("Document")}} value, or `false` otherwise.
- {{domxref("NavigationHistoryEntry.url", "url")}} {{ReadOnlyInline}}
  - : Returns the absolute URL of this history entry.

## Instance methods

_Inherits methods from its parent, {{DOMxRef("EventTarget")}}._

- {{domxref("NavigationHistoryEntry.getState", "getState()")}}
  - : Returns a clone of the available state associated with this history entry.

## Events

- {{domxref("NavigationHistoryEntry/dispose_event", "dispose")}}
  - : Fires when the entry is no longer part of the history entry list.

## Examples

```js

> function initHomeBtn() {
>   // Get the key of the first loaded entry
>   // so the user can always go back to this view.
>   const {key} = navigation.currentEntry;
>   backToHomeButton.onclick = () => {
>     navigation.traverseTo(key);
>   }
> }
> 
> navigation.addEventListener("navigate", event => {
>   event.intercept({
>       async handler() {
>         // Handle single-page navigations
>       }
>   });
> });
> 
> async function handleNavigate(url) {
> 
>   // Navigate to a different view, but the button will always work.
>   await navigation.navigate(url).finished;
> 
>   // ...
> }

Reviewer: Yes, this is clearer (to me anyway). FWIW you don't need this bit:

```js
async function handleNavigate(url) {

  // Navigate to a different view, but the button will always work.
  await navigation.navigate(url).finished;

  // ...
}
```

...because `navigate` will intercept normal links as well, as it does in the demo. You could put the comment `// Navigate to a different view, but the button will always work.` just in the intercept handler:

```suggestion
function initHomeBtn() {
  // Get the key of the first loaded entry
  // so the user can always go back to this view.
  const {key} = navigation.currentEntry;
  backToHomeButton.onclick = () => {
    navigation.traverseTo(key);
  }
}

// Intercept navigate events, such as link clicks, and
// replace them with single-page navigations
navigation.addEventListener("navigate", event => {
  event.intercept({
      async handler() {
        // Navigate to a different view,
        // but the "home" button will always work.
      }
  });
});
```

...just a suggestion though

Author: Good call; I've updated these examples to use your listing suggestion. Thanks Will!
